readme.txt
=== Combined Taxonomies Tag Cloud ===
Contributors:		keith_wp
Author Link:		https://drakard.com/
Requires at least:	3.8 or higher
Tested up to:		5.5
Requires PHP	:	7.2
Stable tag:			0.34
License:			GPLv3 or later
License URI:		https://www.gnu.org/licenses/gpl-3.0.html
Tags:				custom tag cloud, tag cloud, tag, tag cloud widget, change tag cloud

A widget to make tag clouds out of multiple taxonomies, across multiple post types and control their appearance and behavior.


== Description ==

The normal WP Tag Cloud widget only uses one taxonomy at a time, and doesn't properly handle taxonomies being assigned to custom post types.

With this plugin, you can configure (for each cloud):

* which the taxonomies and post types are to be included, and which terms shouldn't be,
* fonts used in the widget and how they scale,
* widget title color and alignment, and general cloud spacing and alignment,
* how the tags look normally, and how they behave when highlighted,
* automatically assign contrasting text colors to meet accessibility guidelines,
* set tags to be highlighted when they have been used elsewhere on that post,
* how the tags should be sorted (alphabetically, by count or just randomly),
* what happens to tags with just one entry (leave alone, remove them or link directly to that post),
* how long each widget's output should be saved for, if at all,
* and now also generate a shortcode for each widget instance and use them within posts.


== Installation ==

1. Upload the plugin files to the **/wp-content/plugins/combined-taxonomies-tag-cloud/** directory, or install the plugin through the WordPress plugins screen directly.
1. Activate the plugin through the 'Plugins' screen in WordPress
1. Use the Appearance -> Widgets screen to add the Combined Tag Cloud widget to your sidebars, customising each widget individually.


== Frequently Asked Questions ==

= How do I select multiple taxonomies or post types? =

Depending on your operating system, you can normally use shift+click to select a range of options in a big dropdown on any webpage, or ctrl+click to select multiple individual ones.

You can also combine them if needed - select loads of options with shift+click first, and then unselect a few individual ones with ctrl+click.

= How do I show a cloud with the shortcode? =

Save an instance of the widget as normal, and the shortcode text will be created at the bottom of the widget form. Copy and paste that snippet where your want the cloud to be displayed - though make sure your theme supports shortcodes there!

If you don't want to display that cloud in a sidebar, then you can either drag it to the "Inactive Widgets" area of the Widgets admin after making it, or create it there directly.


== Screenshots ==

[https://raw.githubusercontent.com/kifd/combined-taxonomies-tag-cloud/master/assets/screenshot-1-types-and-taxonomies.jpg  Pick which taxonomies and post types will be used for this cloud.]

[https://raw.githubusercontent.com/kifd/combined-taxonomies-tag-cloud/master/assets/screenshot-2-fonts.jpg  Alter the font stack and sizes used.]

[https://raw.githubusercontent.com/kifd/combined-taxonomies-tag-cloud/master/assets/screenshot-3-widget-appearance.jpg  Widget title settings and general positioning.]

[https://raw.githubusercontent.com/kifd/combined-taxonomies-tag-cloud/master/assets/screenshot-4-widget-behaviour.jpg  More mechanics of a cloud.]

[https://raw.githubusercontent.com/kifd/combined-taxonomies-tag-cloud/master/assets/screenshot-5-tag-appearance.jpg  Basic appearance of each tag.]

[https://raw.githubusercontent.com/kifd/combined-taxonomies-tag-cloud/master/assets/screenshot-6-tag-effects.jpg  More control over how a tag normally looks and behaves when highlighted.]

[https://raw.githubusercontent.com/kifd/combined-taxonomies-tag-cloud/master/assets/screenshot-7-shortcode.jpg  Use a shortcode to display a cloud in areas outside of widgets.]


== Upgrade Notice ==

= 0.34 =

First WP release since the 0.21.4 version, with a pretty big overhaul of how much the plugin does (see changelog.txt for the list).


changelog.txt
== Changelog ==

= 0.34 =
* Enhancement: added toggle for auto text coloring
* Enhancement: added more choices of effects
* Enhancement: can also change the color of the widget title
* Enhancement: can change the effect transition timing
* Note: update GPL license to 3.0

= 0.33 =
* Enhancement: added tag padding controls
* Enhancement: improved inset shadow, added a few more effects
* Enhancement: tag demo includes more details now (font, border & radius, tag padding)
* Bugfix: transient saves re single page cloud highlighting

= 0.32 =
* Enhancement: changed default wpColorPicker to one that has transparency control
* Enhancement: added more choices of effects
* Enhancement: added demo of tag effect into the widget form
* Enhancement: automatic contrasting text color over backgrounds for a simpler form
* Enhancement: more general widget appearance control (title color, widget padding etc)
* Enhancement: changing taxonomies now automatically updates the possible terms to exclude

= 0.31 =
* Enhancement: font unit selection now reflected across relevant options
* Enhancement: font stack applies to widget title as well
* Enhancement: added choice of tag highlighting effects
* Enhancement: now uses CSS vars for easier customising
* Note: requires php7+ now due to type hinting etc

= 0.30 =
* Enhancement: added ability to display the widget via a shortcode
* Enhancement: added WCAG tag color contast checker
* Enhancement: added option to automatically highlight tags in the cloud that match the tags in the current post
* Enhancement: added adjustable row/column gaps
* Enhancement: can copy the shortcode to the clipboard with a click
* Bugfix: clearing just a color will also tell WP the form needs saving

= 0.23 =
* Enhancement: added 'show count' option
* Enhancement: added tag border color option
* Enhancement: rearranged the admin form UI into collapsible sections
* Enhancement: added mouseover help text to each element
* Enhancement: added tag text decoration option
* Enhancement: completely changed the way CSS was added, so you can style multiple widgets at once
* Bugfix: can now clear color settings, oops
* Bugfix: found a fix for WP not enabling the save button after just a color change

= 0.22 =
* Tested with WP 5.6 alpha
* Enhancement: added tag font stack selection
* Enhancement: added tag text size scaling choice
* Enhancement: added 'vw' to scaling unit selection because why not
* Enhancement: made the widget form hide unused options based on select dropdowns
* Bugfix: checks 'before_widget' param for CSS id, so themes that don't include it (like 2020...) can still have the widgets styled individually
* Bugfix: tidied up the widget form code
* Bugfix: removed a stray pre_get_posts() call

= 0.21.4 =
* Bugfix: was still trying to make a cloud even if we pulled no tags back

= 0.21.3 =
* Bugfix: transient was only saving a single instance
* Bugfix: crashed in a different place if you deleted a taxonomy while it was selected in a widget

= 0.21.2 =
* Tested with WP 5.2.1

= 0.21.1 =
* Tested with WP 4.6.1

= 0.21 =
* Bugfix: crashed if you deleted a taxonomy while it was selected in a widget

= 0.20 =
* Enhancement: added colour pickers
* Bugfix: tidied up the stylesheet

= 0.10 =
* Initial release.


resources/js/admin.js
(function($){
	
	$(document).on('ready widget-added widget-updated', function(event, widget) {
		// WP color picker:
		// thanks to https://www.codecheese.com/2013/04/how-to-use-wp-color-picker-in-widgets/#comment-64600
		// and https://www.codecheese.com/2018/05/trigger-change-wpcolorpicker-in-widgets-and-customizer-page/
		$('.color-picker').not('[id*="__i__"]').wpColorPicker({
			defaultColor: $(this).attr('data-default-color'),
			change: function(e, ui) {
				$(e.target).val(ui.color.to_s('rgba')); // NOTE: not toString anymore; uses the -alpha picker's method
				
				// WP doesn't let you save the widget if all you did was use its own color-picker widget
				$(e.target).trigger('change'); // so trigger manually to enable the save button
			},
			/*clear: function(e) {
				$(e.target).trigger('change'); // likewise if we clear the color
				
				var _id = '#' + $(this).parent().find('.color-picker').attr('id');
				updateColors(_id);
			},*/
		});
		
		// close the fieldsets on initial view, and make sure the extra scaling form fields are shown or not
		$('.combined-taxonomies-tag-cloud legend', widget).each(function() { toggleFieldSet($(this), true); });
		// NOTE: wp resets the widget to its initial state after an update, losing the current open/close state without me
		//       saving them as hidden form fields - possible TODO
		
		
		// NOTE: set the initial states using .trigger() bubbles the event up the dom and makes wp think the
		//       widget needs saving again so I'm calling mini sub functions to work around it
		$('.combined-taxonomies-tag-cloud select.font_unit', widget).each(function() {
			updateFontUnits($(this));
		});
		
		$('*[data-css-var]', widget).each(function() {
			updateCssVars($(this));
		});
		
		
		// hide form elements depending on what has been selected
		$('*[data-controls-others=true]', widget).each(function(idx, dom) {
			toggleManyFields(dom);
		});
		
		
	});
	
	
	// change the font list when updated
	$(document).on('change', 'select.font_family', function() {
		const _stack = (this.value in font_stacks) ? font_stacks[this.value] : '';
		$(this).siblings('.font_list').text(_stack).css('font-family', _stack);
		$(this).parents('form:first').css('--fontFamily', _stack);
	});
	
	
	// change the font units when updated (used for ui feedback and to concatenate onto is-size css-vars)
	$(document).on('change', 'select.font_unit', function() {
		updateFontUnits($(this));
	});
	
	
	// change the vars we use in the tag demo
	$(document).on('change', '*[data-css-var]', function() {
		updateCssVars($(this));
	});
	
	
	// change the tag demo classes when updated
	$(document).on('change', 'select.fx_backgrounds, select.fx_shadows, select.fx_two_dee', function() {
		const widget = $(this).parents('form:first');
		// get the current classes of the element, remove any of the same type (starts with the same fx_?? name) and add the new one
		const element = $('.tag-demo span a', widget);
		const prefix = this.value.substring(0,5);
		let classes = element.attr('class').split(" ").filter(c => !c.startsWith(prefix));
		classes.push(this.value);
		element.attr('class', classes.join(" ").trim());
	});
	
	
	
	$(document).on('change', 'select.taxonomies', function() {
		const widget = $(this).parents('form:first');
		setTerms({
			'widget': widget,
			'selected': $(this).val(),
			'set_dom': $('.exclude', widget),
		});
	});
	
	
	
	
	
	// toggle fieldsets by clicking on the legend
	$(document).on('click', '.combined-taxonomies-tag-cloud legend', function() {
		if ($(this).hasClass('closed')) {
			toggleFieldSet($(this), false);
		} else {
			toggleFieldSet($(this), true);
		}
	});
	
	// toggle other dom elements depending on state of checkboxes/selects
	$(document).on('change', '*[data-controls-others=true]', function() {
		toggleManyFields($(this).get(0));
	});
	
	
	
	
	
	
	// NOTE: units are hard copied from the definition in the .php file
	const font_units = ['rem', 'em', 'pt', 'px', 'vw'];
	let font_unit = 'em';
	
	function updateFontUnits(dom) {
		// TODO: call updateCssVars on data-is-size if this changes
		font_unit = (font_units.indexOf(dom.get(0).value) >= 0) ? dom.get(0).value : 'em';
		const _widget = dom.parents().eq(4);
		
		$('.font_units', _widget).text(font_unit);
	}
	
	
	function updateCssVars(dom) {
		const widget = dom.parents('form:first');
		const cssvar = dom.data('css-var');
		
		let val = dom.val();
		// css vars that need units need them before the browser inserts spaces where it shouldn't
		if (dom.data('is-size'))
			val+= font_unit;
		else if (dom.data('is-time'))
			val+= 's';
		
		$(widget).css('--'+cssvar, val);
		console.log('set '+cssvar+' to '+val);
		
		
		const wcag = {
			'backColor1': 'textColor1',
			'backColor2': 'textColor2',
			//'widgetBackgroundColor': both
		};
		
		
		if (wcag.hasOwnProperty(cssvar) && $('.auto_text_color', widget).prop('checked')) {
			
			let color = {
				'against': dom.val(),
				'widget': $(widget),
				'set_var': wcag[cssvar],
				'wcag': $(dom).parents('p').children('.wcag'),
				'additional': $(widget).css('--widgetBackgroundColor')
			}
			
			// NOTE: if the widget background is transparent, then we assume the background behind that is white
			// TODO: implement dark mode or use magic to find out what it really is
			if (typeof color.additional === 'undefined') color.additional = 'rgba(255,255,255,1)';
			
			// TODO: replace the ajax call to php with js implementation
			setContrast(color);
		}
	}

	
	
	
	// like toggleClass except won't open a closed one if you didn't want it to 
	function toggleFieldSet(dom, close = true) {
		if (close && ! dom.hasClass('closed')) {
			dom.addClass('closed');
		} else if (! close && dom.hasClass('closed')) {
			dom.removeClass('closed');
		}
	}
	
	
	// marginally more advanced version of the original toggle, certainly more complex
	function toggleManyFields(dom) {
		
		// NOTE: avoid having both attributes on the same element without checking the order
		// NOTE: also, different elements could have conflicting hide/show demands - again, check the order or redesign the form
		let toggles = { 'show': [], 'hide': [] };
		const widget = $(dom).parents('form:first');
		
		if (dom.nodeName == 'INPUT' && dom.type == 'checkbox') {
			
			if (typeof $(dom).data('hide-these') !== 'undefined') {
				const state = (dom.checked == true) ? 'hide' : 'show';
				const elements = $(dom).data('hide-these').split(',');
				
				$(elements).each(function(i, which_string) {
					// takes advantage of my form being made of "p > label + thing" blocks
					toggles[state].push($('[id*="'+which_string+'"]', widget).parents('p:first'));
				});
			}
			if (typeof $(dom).data('show-these') !== 'undefined') {
				const state = (dom.checked == true) ? 'show' : 'hide';
				const elements = $(dom).data('show-these').split(',');
			
				$(elements).each(function(i, which_string) {
					toggles[state].push($('[id*="'+which_string+'"]', widget).parents('p:first'));
				});
			}
		
		} else if (dom.nodeName == 'SELECT')  {
			// selects are awkward
			
			$(dom).children('option').each(function(i, option) {
				if (typeof $(option).data('hide-these') !== 'undefined' && option.selected == true) {
					$($(option).data('hide-these').split(',')).each(function(i, which_string) {
						toggles['hide'].push($('[id*="'+which_string+'"]', widget).parents('p:first'));
					});
				}
				if (typeof $(option).data('show-these') !== 'undefined' && option.selected == true) {
					$($(option).data('show-these').split(',')).each(function(i, which_string) {
						toggles['show'].push($('[id*="'+which_string+'"]', widget).parents('p:first'));
					});
				}
				// update the contrast demo if needed
				if (typeof $(option).data('contrast') !== 'undefined' && option.selected == true) {
					$(option).parents('p:first').nextAll('p.color-demo').attr('data-contrast', $(option).data('contrast'));
				}
			});
		}
		
		$(toggles.hide).each(function(i, what) {
			what.addClass('hide');
			$('input', what).each(function(j, dom) {
				if (typeof $(dom).data('css-var') !== 'undefined') {
					$(widget).css('--'+$(dom).data('css-var'), '');
				}
			});
		});
		$(toggles.show).each(function(i, what) {
			what.removeClass('hide');
			$('input', what).each(function(j, dom) {
				if (typeof $(dom).data('css-var') !== 'undefined') {
					updateCssVars($(dom));
				}
			});
		});
	}
	
	
	
	function setContrast(color) {
		$.ajax({
			type:			'POST',
			url:			cttc_ajax.url,
			data:			{
								'action': 'cttc_get_contrast',
								'_ajax_nonce': cttc_ajax.nonce,
								'color': color.against,			// the background we're contrasting against
								'additional': color.additional, // and the widget background to the background in case of transparency
							},
			success:		function(response) {
								if (response.ok == true) {
									$(color.widget).css('--'+color.set_var, response.contrast);
									$(color.wcag).html(response.wcag);
									
									console.log('cool', response, color);
									//console.log(color.against, response.contrast, response.ratio, response.wcag);
									
								} else {
									console.log('error', response, color);
									//$(color.wcag).html('Error');
								}
							},
			dataType:		'json',
		});
	}
	
	
	function setTerms(taxonomies) {
		$.ajax({
			type:			'POST',
			url:			cttc_ajax.url,
			data:			{
								'action': 'cttc_get_terms',
								'_ajax_nonce': cttc_ajax.nonce,
								'taxonomies': taxonomies.selected,
							},
			success:		function(response) {
								if (response.ok == true) {
									$(taxonomies.set_dom).empty().html(response.options);
									/*$.each(response.terms, function(k, v) {
										$(taxonomies.set_dom).append("<option value='"+v.term_id+"'>"+v.name+'</option>');
									});*/
									//console.log('cool', response);
									
								} else {
									console.log('error', response);
								}
							},
			dataType:		'json',
		});
	}
	
})(jQuery);



// https://www.jasongaylord.com/blog/2020/05/21/copy-to-clipboard-using-javascript
var copy_elements = document.querySelectorAll("[data-copy-text]");
var copy_count;
for (copy_count = 0; copy_count < copy_elements.length; copy_count++) {
	copy_elements[copy_count].addEventListener("click", async(event) => {
		if (! navigator.clipboard) {
			return;
		}
		try {
			var copy_value = event.srcElement.getAttribute("data-copy-text");
			await navigator.clipboard.writeText(copy_value);
		} catch (error) {
			console.error("copy failed", error);
		}
	});
}


combined-taxonomies-tag-cloud.php
<?php
/*
Plugin Name:  Combined Taxonomies Tag Cloud
Version:      0.34
Description:  Makes a tag cloud widget out of multiple taxonomies across multiple post types.
Author:       Keith Drakard
Author URI:   https://drakard.com/
Requires PHP: 7.0
Text Domain:  CombinedTaxonomiesTagCloud
Domain Path:  /languages
*/


if (! defined('WPINC')) die;

class CombinedTaxonomiesTagCloudPlugin {
	
	public function __construct() {
		load_plugin_textdomain('CombinedTaxonomiesTagCloud', false, dirname(plugin_basename(__FILE__)).'/languages');
		
		add_action('widgets_init', function() {
			require_once 'cttc-widget.php';
			register_widget('CombinedTaxonomiesTagCloudWidget');
			// unregister_widget('WP_Widget_Tag_Cloud');
		});
		
		add_shortcode('cttc', array($this, 'widget_shortcode'));
	}
	
	
	// only previously saved widgets can be called like this, else I'm duplicating the entire params
	public function widget_shortcode($attributes, $content = null) {
		$defaults = array(
			'cloud'	=> 0,
			'echo'	=> false,
		);
		$args = shortcode_atts($defaults, $attributes, 'cttc');
		
		// ensure we have the same types as expected from the defaults
		foreach ($args as $key => $value) settype($args[$key], gettype($defaults[$key]));
		
		$output = '';
		if ($args['cloud'] > 0) {
			
			// all our widget instances are saved in the options table
			$widget_options_all = get_option('widget_combinedtaxonomiestagcloudwidget');
			
			if (is_array($widget_options_all) AND isset($widget_options_all[$args['cloud']])) {
				$instance = $widget_options_all[$args['cloud']];
				// add the id of this widget, doctored to show it's via the shortcode
				$instance['widget_id'] = 'combinedtaxonomiestagcloudwidget-'.$args['cloud'].'-shortcode';
				// get the widget output
				ob_start(); the_widget('CombinedTaxonomiesTagCloudWidget', $instance); $output = ob_get_clean();
				
			} else {
				$output = sprintf('<p>%s</p>',
					sprintf(__('CTTC Error: "%s" was not recognized as a valid tag cloud ID', 'CombinedTaxonomiesTagCloud'), $attributes['cloud'])
				);
			}
		} elseif (isset($attributes['cloud'])) {
			$output = sprintf('<p>%s</p>',
				sprintf(__('CTTC Error: "%s" was not recognized as a numeric tag cloud ID', 'CombinedTaxonomiesTagCloud'), $attributes['cloud'])
			);
		} else {
			$output = sprintf('<p>%s</p>',
				sprintf(__('CTTC Error: You need to enter a tag cloud ID to show a tag cloud', 'CombinedTaxonomiesTagCloud'))
			);
		}
		
		if ($args['echo']) {
			echo $output;
		}
		
		return $output;
	}
	
}

$CombinedTaxonomiesTagCloud = new CombinedTaxonomiesTagCloudPlugin();



config.php
<?php

$config = array();

$config['fonts'] = array(
	// from https://gist.github.com/don1138/5761014
	'System' => 
		'system, -apple-system, ".SFNSText-Regular", "San Francisco", "Roboto", "Segoe UI", "Helvetica Neue", "Lucida Grande", sans-serif',
	'Times New Roman' =>
		'Cambria, "Hoefler Text", Utopia, "Liberation Serif", "Nimbus Roman No9 L Regular", Times, "Times New Roman", serif',
	'Georgia' =>
		'Constantia, "Lucida Bright", Lucidabright, "Lucida Serif", Lucida, "DejaVu Serif", "Bitstream Vera Serif", "Liberation Serif", Georgia, serif',
	'Garamond' =>
		'"Palatino Linotype", Palatino, Palladio, "URW Palladio L", "Book Antiqua", Baskerville, "Bookman Old Style", "Bitstream Charter", "Nimbus Roman No9 L", Garamond, "Apple Garamond", "ITC Garamond Narrow", "New Century Schoolbook", "Century Schoolbook", "Century Schoolbook L", Georgia, serif',
	'Helvetica/Arial' =>
		'Frutiger, "Frutiger Linotype", Univers, Calibri, "Gill Sans", "Gill Sans MT", "Myriad Pro", Myriad, "DejaVu Sans Condensed", "Liberation Sans", "Nimbus Sans L", Tahoma, Geneva, "Helvetica Neue", Helvetica, Arial, sans-serif',
	'Verdana' =>
		'Corbel, "Lucida Grande", "Lucida Sans Unicode", "Lucida Sans", "DejaVu Sans", "Bitstream Vera Sans", "Liberation Sans", Verdana, "Verdana Ref", sans-serif',
	'Trebuchet' =>
		'"Segoe UI", Candara, "Bitstream Vera Sans", "DejaVu Sans", "Bitstream Vera Sans", "Trebuchet MS", Verdana, "Verdana Ref", sans-serif',
	'Impact' =>
		'Impact, Haettenschweiler, "Franklin Gothic Bold", Charcoal, "Helvetica Inserat", "Bitstream Vera Sans Bold", "Arial Black", sans-serif',
	'Monospace 1' =>
		'Consolas, "Andale Mono WT", "Andale Mono", "Lucida Console", "Lucida Sans Typewriter", "DejaVu Sans Mono", "Bitstream Vera Sans Mono", "Liberation Mono", "Nimbus Mono L", Monaco, "Courier New", Courier, monospace',
		
	// https://www.webfx.com/blog/web-design/a-web-designers-guide-to-linux-fonts/
	'Narrow Sans-Serif' =>
		'"Liberation Sans", "Nimbus Sans L", "FreeSans", "Helvetica Neue", Helvetica, Arial, sans-serif',
	'Wide Sans-Serif' =>
		'"DejaVu Sans", "Bitstream Vera Sans", Geneva, Verdana, sans-serif',
	'Narrow Serif' =>
		'"Liberation Serif", "Nimbus Roman No 9 L",  "FreeSerif", "Hoefler Text", Times, "Times New Roman", serif',
	'Antique Serif' =>
		'"Bitstream Charter", "URW Palladio L", Palatino, "Palatino Linotype", "Book Antiqua", serif',
	'Wide Serif' =>
		'"DejaVu Serif", "Bitstream Vera Serif", "Century Schoolbook L", "Lucida Bright", Georgia, serif',
	'Monospace 2' =>
		'"Liberation Mono", "Nimbus Mono L", "FreeMono", "DejaVu Mono", "Bitstream Vera Mono", "Lucida Console", "Andale Mono", "Courier New", monospace',
	
	
);
ksort($config['fonts']);


// NOTE: $fields var in form() function must have matching keys
// TODO: given that I've now expanded the values for the show/hide controls, add the rest of the field definitions and be done with it...
$config['choices'] = array(
	'align_h'			=> array(
							'left'			=> __('Left', 'CombinedTaxonomiesTagCloud'),
							'center'		=> __('Center', 'CombinedTaxonomiesTagCloud'),
							'right'			=> __('Right', 'CombinedTaxonomiesTagCloud'),
						),
	'align_title'		=> array(
							'left'			=> __('Left', 'CombinedTaxonomiesTagCloud'),
							'center'		=> __('Center', 'CombinedTaxonomiesTagCloud'),
							'right'			=> __('Right', 'CombinedTaxonomiesTagCloud'),
						),
	'align_v'			=> array(
							'start'			=> __('Top', 'CombinedTaxonomiesTagCloud'),
							'center'		=> __('Center', 'CombinedTaxonomiesTagCloud'),
							'end'			=> __('Bottom', 'CombinedTaxonomiesTagCloud'),
						),
	'border_style'		=> array(
							''				=> array(
								'name'			=> __('None', 'CombinedTaxonomiesTagCloud'),
								'hide-these'	=> array('tborder1', 'tborder2', 'border_width'),
							),
							'solid'			=> array(
								'name'			=> __('Solid', 'CombinedTaxonomiesTagCloud'),
								'show-these'	=> array('tborder1', 'tborder2', 'border_width'),
							),
							'dotted'		=> array(
								'name'			=> __('Dots', 'CombinedTaxonomiesTagCloud'),
								'show-these'	=> array('tborder1', 'tborder2', 'border_width'),
							),
							'dashed'		=> array(
								'name'			=> __('Dashes', 'CombinedTaxonomiesTagCloud'),
								'show-these'	=> array('tborder1', 'tborder2', 'border_width'),
							),
							'double'		=> array(
								'name'			=> __('Double Lines', 'CombinedTaxonomiesTagCloud'),
								'show-these'	=> array('tborder1', 'tborder2', 'border_width'),
							),
							'groove'		=> array(
								'name'			=> __('Grooved', 'CombinedTaxonomiesTagCloud'),
								'show-these'	=> array('tborder1', 'tborder2', 'border_width'),
							),
							'ridge'			=> array(
								'name'			=> __('Ridged', 'CombinedTaxonomiesTagCloud'),
								'show-these'	=> array('tborder1', 'tborder2', 'border_width'),
							),
							'inset'			=> array(
								'name'			=> __('Inset', 'CombinedTaxonomiesTagCloud'),
								'show-these'	=> array('tborder1', 'tborder2', 'border_width'),
							),
							'outset'		=> array(
								'name'			=> __('Outset', 'CombinedTaxonomiesTagCloud'),
								'show-these'	=> array('tborder1', 'tborder2', 'border_width'),
							),
						),
	
	'fx_backgrounds'	=> array(
							'fx_bg_none' 		=> array(
								'name'				=> __('None', 'CombinedTaxonomiesTagCloud'),
								'hide-these'		=> array('tcolor1', 'tcolor2'),
							),
							'fx_bg_static' 		=> array(
								'name'				=> __('Static', 'CombinedTaxonomiesTagCloud'),
								'show-these'		=> array('tcolor1'),
								'hide-these'		=> array('tcolor2'),
							),
							'fx_bg_fade'		=> array(
								'name'				=> __('Switch', 'CombinedTaxonomiesTagCloud'),
								'show-these'		=> array('tcolor1', 'tcolor2'),
							),
							'fx_bg_sweep_up'	=>  array(
								'name'				=> __('Sweep Up', 'CombinedTaxonomiesTagCloud'),
								'show-these'		=> array('tcolor1', 'tcolor2'),
							),
							'fx_bg_sweep_down'	=>  array(
								'name'				=> __('Sweep Down', 'CombinedTaxonomiesTagCloud'),
								'show-these'		=> array('tcolor1', 'tcolor2'),
							),
							'fx_bg_sweep_left'	=>  array(
								'name'				=> __('Sweep Left', 'CombinedTaxonomiesTagCloud'),
								'show-these'		=> array('tcolor1', 'tcolor2'),
							),
							'fx_bg_sweep_right'	=>  array(
								'name'				=> __('Sweep Right', 'CombinedTaxonomiesTagCloud'),
								'show-these'		=> array('tcolor1', 'tcolor2'),
							),
							'fx_bg_radial_in'	=>  array(
								'name'				=> __('Radial In', 'CombinedTaxonomiesTagCloud'),
								'show-these'		=> array('tcolor1', 'tcolor2'),
							),
							'fx_bg_radial_out'	=>  array(
								'name'				=> __('Radial Out', 'CombinedTaxonomiesTagCloud'),
								'show-these'		=> array('tcolor1', 'tcolor2'),
							),
							'fx_bg_rectangle_in'	=>  array(
								'name'				=> __('Rectangle In', 'CombinedTaxonomiesTagCloud'),
								'show-these'		=> array('tcolor1', 'tcolor2'),
							),
							'fx_bg_rectangle_out'	=>  array(
								'name'				=> __('Rectangle Out', 'CombinedTaxonomiesTagCloud'),
								'show-these'		=> array('tcolor1', 'tcolor2'),
							),
							'fx_bg_underline_center_out'	=> array(
								'name'				=> __('Underline Center Out', 'CombinedTaxonomiesTagCloud'),
								'show-these'		=> array('tcolor1', 'tcolor2'),
							),
							'fx_bg_overline_center_out'	=> array(
								'name'				=> __('Overline Center Out', 'CombinedTaxonomiesTagCloud'),
								'show-these'		=> array('tcolor1', 'tcolor2'),
							),
							'fx_bg_bothline_center_out'	=> array(
								'name'				=> __('Both Center Out', 'CombinedTaxonomiesTagCloud'),
								'show-these'		=> array('tcolor1', 'tcolor2'),
							),
						),
	'fx_shadows'		=> array(
							'fx_sh_none' 			=> array(
								'name'				=> __('None', 'CombinedTaxonomiesTagCloud'),
								'hide-these'		=> array('tshadow'),
							),
							'fx_sh_shadow'			=> array(
								'name'				=> __('Shadow', 'CombinedTaxonomiesTagCloud'),
								'show-these'		=> array('tshadow'),
							),
							'fx_sh_inset'			=> array(
								'name'				=> __('Shadow Inset', 'CombinedTaxonomiesTagCloud'),
								'show-these'		=> array('tshadow'),
							),
							'fx_sh_glow'			=> array(
								'name'				=> __('Shadow Glow', 'CombinedTaxonomiesTagCloud'),
								'show-these'		=> array('tshadow'),
							),
							'fx_sh_box_inset'		=> array(
								'name'				=> __('Box Shadow Inset', 'CombinedTaxonomiesTagCloud'),
								'show-these'		=> array('tshadow'),
							),
							'fx_sh_box_outset'		=> array(
								'name'				=> __('Box Shadow Outset', 'CombinedTaxonomiesTagCloud'),
								'show-these'		=> array('tshadow'),
							),
						),
	'fx_two_dee'		=> array(
							'fx_2d_none' 		=> array(
								'name'				=> __('None', 'CombinedTaxonomiesTagCloud'),
							),
							'fx_2d_grow'		=> array(
								'name'				=> __('Grow', 'CombinedTaxonomiesTagCloud'),
							),
							'fx_2d_shrink'		=> array(
								'name'				=> __('Shrink', 'CombinedTaxonomiesTagCloud'),
							),
							'fx_2d_rotate'		=> array(
								'name'				=> __('Rotate', 'CombinedTaxonomiesTagCloud'),
							),
							'fx_2d_grow_rotate'		=> array(
								'name'				=> __('Grow &amp; Rotate', 'CombinedTaxonomiesTagCloud'),
							),
							'fx_2d_skew'		=> array(
								'name'				=> __('Skew', 'CombinedTaxonomiesTagCloud'),
							),
						),
	
	'font_family'		=> array(__('Leave Alone', 'CombinedTaxonomiesTagCloud')),
	'font_unit'			=> array('rem', 'em', 'pt', 'px', 'vw'),
	'highlight'			=> array(
							''				=> __('None', 'CombinedTaxonomiesTagCloud'),
							'single'		=> __('On Single Pages', 'CombinedTaxonomiesTagCloud'),
						),
	'orderby'			=> array(
							'name'			=> __('Alphabetically', 'CombinedTaxonomiesTagCloud'),
							'count'			=> __('By Count', 'CombinedTaxonomiesTagCloud'),
							'random'		=> __('Randomly', 'CombinedTaxonomiesTagCloud')
						),
	'post_types'		=> get_post_types(array('show_ui' => true), 'objects'),
	'save'				=> array(0, 1, 2, 4, 8, 12, 24, 48, 96), // hours
	'single'			=> array(
							''				=> __('Leave Alone', 'CombinedTaxonomiesTagCloud'),
							'remove'		=> __('Remove', 'CombinedTaxonomiesTagCloud'),
							'link'			=> __('Link to Entry', 'CombinedTaxonomiesTagCloud')
						),
	'taxonomies'		=> get_taxonomies(array('show_ui' => true), 'objects'),
	'text_case'			=> array(
							''				=> __('Leave Alone', 'CombinedTaxonomiesTagCloud'),
							'lower'			=> __('lowercase', 'CombinedTaxonomiesTagCloud'),
							'upper'			=> __('UPPERCASE', 'CombinedTaxonomiesTagCloud')
						),
	'text_decoration'	=> array(
							''				=> __('Leave Alone', 'CombinedTaxonomiesTagCloud'),
							'no_text_deco'	=> __('None', 'CombinedTaxonomiesTagCloud'),
							'under_always'	=> __('Always Underline', 'CombinedTaxonomiesTagCloud'),
							'under_hover'	=> __('Underline on Hover', 'CombinedTaxonomiesTagCloud'),
							'box_hover'		=> __('Box Shadow on Hover', 'CombinedTaxonomiesTagCloud'),
							'under_dotted'	=> __('Underline to Dots', 'CombinedTaxonomiesTagCloud'),
						),
);
sort($config['choices']['taxonomies']);
sort($config['choices']['post_types']);

$config['defaults'] = array(
	'align_h'			=> 'left',
	'align_title'		=> 'left',
	'align_v'			=> 'start',
	'auto_text_color'	=> 1,
	'border_radius'		=> 0.10,
	'border_style'		=> '',
	'border_width'		=> 0.01,
	'column_gap'		=> 0.50,
	'exclude'			=> array(0),
	'font_base'			=> 1.00,
	'font_family'		=> __('Leave Alone', 'CombinedTaxonomiesTagCloud'),
	'font_unit'			=> 'em',
	'fx_backgrounds'	=> 'fx_bg_none',
	'fx_shadows'		=> 'fx_sh_none',
	'fx_two_dee'		=> 'fx_2d_none',
	'fx_timing'			=> 0.3,
	'highlight'			=> '',
	'largest' 			=> 1.40,
	'maximum'			=> 999,
	'nofollow'			=> 0,
	'order'				=> 0, // 0 = asc, 1 = desc
	'orderby'			=> 'name',
	'post_types'		=> array('post'),
	'row_gap'			=> 0.50,
	'save'				=> 0,
	'scale_tag'			=> 0,
	'show_count'		=> 1,
	'single'			=> '',
	'smallest' 			=> 0.60,
	'taxonomies'		=> array('post_tag'),
	'tcolor1'			=> [255,255,255,0],
	'tcolor2'			=> [255,255,255,0],
	'tborder1'			=> '#ffffffff',
	'tborder2'			=> '#000000ff',
	'tag_padding_x'		=> 1.00,
	'tag_padding_y'		=> 0.50,
	'tag_text_color_1'	=> [0,0,0,1],
	'tag_text_color_2'	=> [0,0,0,1],
	'tshadow'			=> '#00000099',
	'text_case'			=> '',
	'text_decoration'	=> 'under_hover',
	'title' 			=> '',
	'title_color'		=> [0,0,0,1],
	'wbackground'		=> [255,255,255,0],
	'wborder_radius'	=> 0.00,
	'wpadding'			=> 0.00,
);
